### ISSUE TITLE
Access RLMResults from different thread fails

### ISSUE THREAD

**Issue by prgorasiya** [realm/realm-cocoa#3936]
I am using Realm Objective C into my application and I really like it because of all its features, easy config and speed.

Currently I am facing an issue but I don't know if thats already been fixed in latest version of Realm objective C because I haven't updated my pods because of some other issues.

The issue is I am initialising an RLMResult object onto main thread and filling it into background thread and after that whenever needed I am accessing it onto main thread and my app crashes at that time.

So, I want to know if I can achieve such functionality or is there any other way around ?

Realm version: 0.98.6

Xcode version: 7.3.1

iOS/OSX version: 9.3

Dependency manager + version: Cocoa pods 0.39

Thanks in advance

**Comment by bdash** [comment]
Can you clarify why you're trying to access `RLMResults` from multiple threads? Are you trying to perform a query in the background and then access the results from the main thread?

**Comment by prgorasiya** [comment]
I am loading my form and it requires some different datasource so I am using `RLMResults` for that.
The form takes some time when loading so I wanted to fill all those `RLMResults` objects into background thread so I can load the form quickly and access datasource lately.
So, when I try to access `RLMResults` object after loading my view, my app crashes saying Realm accessed from different thread.
Basically I am filing all the `RLMResults` objects into background thread and trying to access those in main thread.
Note that I have taken all the `RLMResults` objects global in my class.

**Comment by Kirow** [comment]
You can access realm object only from different threads, [see doc](https://realm.io/docs/objc/latest/#threads). 

But I think [Realm notifications](https://realm.io/docs/objc/latest/#realm-notifications) will work the way you want. It will execute your query asynchronously and throw result into main thread.

**Comment by bdash** [comment]
Yes, as @Kirow mentions notifications seem like the solution for this problem. In particular, [Collection Notifications](https://realm.io/docs/objc/latest/#collection-notifications) allow you to perform a query that will evaluate its results in a background thread and then provides them to your main thread. You will also be notified when the results of the query change in the future so you can keep your UI up to date.

**Comment by prgorasiya** [comment]
I know about notifications but that's not satisfying my requirements.

I am doing following operations on main thread:

```
self.industryArray = [[NSMutableArray alloc] init];
self.industryIndexArray = [[NSMutableArray alloc] init];

GlobalPicklist *globalPicklistObject = [[GlobalPicklist alloc] init];

self.industry = [[GlobalPicklist objectsWhere:@"globalPicklistCategoryIndex = %@",IndexForIndustryPicklist] sortedResultsUsingProperty:@"globalPicklistSortValue" ascending:YES];

for (int i=0; i<[self.industry count]; i++)
    {
        
        globalPicklistObject = [self.industry objectAtIndex:i];
        [self.industryArray addObject:globalPicklistObject.globalPicklistName];
        [self.industryIndexArray addObject:globalPicklistObject.globalPicklistIndex];
    }
```

But I want to perform on background thread since there 4 more loops that I require.

And I am accessing them onto main thread by:

```
NSMutableArray *industryNameArray = [[NSMutableArray alloc] init];
        
        for (int i=0; i<[[[NSUserDefaults standardUserDefaults] objectForKey:@"addCompanyIndustryList"] count]; i++)
        {
            globalObject = [[self.industry objectsWhere:@"globalPicklistIndex = %@",[[[NSUserDefaults standardUserDefaults] objectForKey:@"addCompanyIndustryList"] objectAtIndex:i]] firstObject];
            [industryNameArray addObject:globalObject.globalPicklistName];
        }
```

So, can help me with this scenario to support background operations by any other way?

**Comment by jpsim** [comment]
We can't take the code you shared in your last comment and adapt it to work across threads, because it's explicitly violating Realm's thread confinement rules. However, your initial comment shares what you're actually trying to accomplish, and that can certainly be done with the [collection notifications](https://realm.io/docs/objc/latest/#collection-notifications) that @bdash suggested above.

> The issue is I am initialising an RLMResult object onto main thread and filling it into background thread and after that whenever needed I am accessing it onto main thread and my app crashes at that time.

So instead of placing Realm objects obtained from a background thread in an `NSMutableArray` to be accessed on the main thread, you should use this pattern:

```objc
// subscribe to collection notifications on main thread
self.notificationToken = [self.industries addNotificationBlock:^(RLMArray<Industry *> *industries, RLMCollectionChange *changes, NSError *error) {
  // deal with updated values
}];

// write on background queue:
dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
  RLMRealm *realm = [RLMRealm defaultRealm];
  [realm beginWriteTransaction];
  // add items to an RLMArray, maybe in your GlobalPicklist? This depends on your models.
  [realm commitWriteTransaction];
});
```

**Comment by prgorasiya** [comment]
@jpsim I think you misunderstood my question. I am not trying to add values to database model, I am trying to fill RLMResult object from my database model, so there is no need for Realm transaction.
And I don't need collection notification in particularly my case because there no need to update UI after RLMResult objects has been filled.
What I want to achieve is when my RLMResult object is filled I want that object as a datasource option for one of picker in my form and I want to fill all those RLMResult objects into background so my form loads quickly.

Thanks

**Comment by jpsim** [comment]
> I am trying to fill RLMResult object from my database model, so there is no need for Realm transaction.

`RLMResults` are entirely read-only. They're views into the underlying data, usually backed by queries. So the only way to modify them (e.g. "fill" them) would be to make Realm transactions, which is why @bdash and I have proposed the solutions mentioned above.

I strongly encourage you to read our documentation on [Writes](https://realm.io/docs/objc/latest/#writes), [Queries](https://realm.io/docs/objc/latest/#queries) and [RLMResults](https://realm.io/docs/objc/latest/api/Classes/RLMResults.html).

**Comment by prgorasiya** [comment]
I totally understand that RLResults are read-only and I am not trying to modify them either. I am just accessing them from different thread.
What I want to do is fill them once in background thread and access them from main thread.
And I think I am not able to understand how in my case Collection notifications will be useful.
Please provide some details on that.

thanks

**Comment by jpsim** [comment]
Collection notifications let you know when you collection have changed. An RLMResults is a Realm collection. This is explained in the [Collections](https://realm.io/docs/objc/latest/#collections) section of the docs.

The reason you'd want to use collection notifications is to know when the changes you made in your background write transactions have propagated to the main thread.

**Comment by jpsim** [comment]
Closing since we haven't heard back from @prgorasiya for a few weeks.
